modin/experimental/batch/pipeline.py

```diff
@@ -0,0 +1,354 @@
+            )
+            return []
+        if self.node_to_id is None and pass_output_id:
+            raise ValueError(
+                "`pass_output_id` is set to True, but output ids have not been specified. "
+                + "To pass output ids, please specify them using the `output_id` kwarg with pipeline.add_query"
+            )
+        if self.node_to_id:
+            outs = {}
+        else:
+            outs = []
+        modin_frame = self.df._query_compiler._modin_frame
+        ptns = modin_frame._partition_mgr_cls.row_partitions(modin_frame._partitions)
+        for i, node in enumerate(self.outputs):
+            ptns = self._complete_nodes(node.operators + [node], ptns)
+            for ptn in ptns:
+                ptn.drain_call_queue(num_splits=1)
+            out_ptns = ptns
+            if postprocessor:
+                args = []
+                if pass_output_id:
+                    args.append(self.node_to_id[node])
+                if pass_partition_id:
+                    args.append(0)
+                    new_ptns = []
+                    for i, p in enumerate(ptns):
+                        new_ptns.append(p.add_to_apply_calls(postprocessor, *args))
+                        args[-1] = i + 1
+                    out_ptns = new_ptns
+                else:
+                    out_ptns = [
+                        ptn.add_to_apply_calls(postprocessor, *args) for ptn in ptns
+                    ]
+                if final_result_func is None:
+                    [ptn.drain_call_queue(num_splits=1) for ptn in out_ptns]
+            if self.node_to_id is None:
+                outs.append(out_ptns)
+            else:
+                outs[self.node_to_id[node]] = out_ptns
+        wait_parts = []
+        if self.node_to_id is None:
+            final_results = []
+            for df in outs:
+                if final_result_func:
+                    d = df[0].to_pandas()
+                    final_results.append(final_result_func(d))
+                else:
+                    ptns = np.array(df).flatten()
+                    ptns = np.array([ptn.list_of_blocks for ptn in df]).flatten()
+                    final_results.append(from_partitions(ptns, Axis.ROW_WISE.value))
+                for d in df:
+                    wait_parts.extend(d.list_of_blocks)
+        else:
+            final_results = {}
+            for id, df in outs.items():
+                if final_result_func:
+                    d = df[0].to_pandas()
+                    final_results[id] = final_result_func(d)
+                else:
+                    ptns = np.array(df).flatten()
+                    ptns = np.array([ptn.list_of_blocks for ptn in df]).flatten()
+                    final_results[id] = from_partitions(ptns, Axis.ROW_WISE.value)
+                for d in df:
+                    wait_parts.extend(d.list_of_blocks)
+        ray.wait(wait_parts, num_returns=len(wait_parts))
```

Comment:
`wait_parts` is the output. We need to wait for it as otherwise we won't fully compute the outputs.

Reply:
Nevertheless, why should we wait for `wait_parts`? Users will be able to materialize the data when needed by calling `print` or any other related stuff?

Reply:
I personally prefer materializing it, because I think that `compute_batch` should present the final materialized results to the user (e.g. if they pass this to a .csv or to some other process afterwards) since this is a batched workload and not an interactive one. @devin-petersohn thoughts? I can change it but that would probably require users to put a call in that forces materialization at the end of the pipeline, which is extra work for the user.

Reply:
I do not see a problem. Users will be dealing with modin.pandas.DataFrame object that hides all internals (partitions).

Reply:
@RehanSD, is this addressed?

Reply:
Same here - we no longer do any waiting. We do split the final result back up into block partitions though.